On-Chip Debugging for AVR/ATmega based boards

maribu opened the pull request:
### Contribution description
 - Refactor/deduplicate `debug.sh`, `debug_srv.sh` and `gdb.conf` of the ATmega based boards
    - This fixes the broken `make debug` for Arduino Mega2560
- Allow to specify the hardware (debugger device) via the `DEBUGDEVICE` environment variable, instead of sticking to the device the developer of the specific board used.
    - Use the AVR Dragon by default, which is the ~~cheapest~~ least expensive programmer&debugger (thus hacker friendly) and supports the most AVR MCUs including all supported by RIOT (thus a good default for every ATmega board).
- Choose the debug protocol used based on the MCU.
    - Currently JTAG will be used unless the ATmega328P MCU is used, which supports debugWIRE instead of JTAG
- Added documentation on how to do On-Chip Debugging for the Arduino Uno
- Fixed incorrect flags for `avrdude` for the ATmega based Arduinos
     - Previously signature verification was disabled, which allows users to mistakenly flash an image intended for a completely different MCU. This is dangerous and should only be enabled by users knowing what the do for very rare and specific reasons
    - Previously auto erase was disabled. Citing the avrdude doc "it generally is a mistake to program the flash without performing an erase first" and results in corrupted flash contents when an ISP is used.
        - Added disabled of auto-erase specifically for the Arduino Mega2560 if (and only if) the default programmer is used, to work around a specific bug in the bootloader

### Testing procedure
Beware: You will need an AVR Dragon for testing. (At least for the Arduino Uno, you can test the Arduino Mega2560 with e.g. an Atmel ICE as well.)

- [ ] Arduino Uno:
    - Try to do some on-chip debugging. I'm not explaining how to do that here, so that you will test the provided doc by doing so as well
- [ ] Arduino Mega2560:
    - Try to do on-chip debugging. Read the doc (which is already in master) on how to do so. It should now work again
- [ ] Some other AVR based board ideally with a different debugger than the AVR Dragon

### Issues/PRs references
None

------

Update: Adapted description

kYc0o commented:
Can you take a look to #9130 ? I think this PR makes more sense on top of that one.

maribu commented:
I suggest I'll rebase on top of https://github.com/RIOT-OS/RIOT/pull/9130 once it is merged and this PR remains open until then

kYc0o commented:
OK, perfect! Thanks!

maribu commented:
OK, I just re-visited this PR and basically rewrote it, as the changes no longer applied to RIOT due to a lot of refactoring. I was so free to rebase and force push to solve the merge conflicts.

maribu commented:
I extended the doc a bit. Especially the point about how the Arduino Uno needs to be modified in order to be able to use debugWIRE was quite important to write and somehow I still managed to forget.

aabadie commented:
I can give this a try on arduino-mega2560.

Note that this needs a rebase.

maribu commented:
Do you have an AVR-Dragon? I could bring mine to the summit, so that you could borrow it for testing.

aabadie commented on boards/arduino-uno/doc.txt:
```diff
@@ -37,6 +37,119 @@ This should take care of everything!
+You can enable debugWIRE debugging using the ISP of the AVR Dragon by
+running:
+
+    avrude -p m328p -c dragon_isp -U lfuse:w:0xff:m -U hfuse:w:0x9e:m -U efuse:w:0xfd:m
```
Should be avr**d**ude, here and below

maribu commented:
Rebased to fix the merge conflict. (Apart from the manual merge of `boards/arduino-uno/doc.text`, the commits of this PR should be unchanged.)

maribu commented on boards/arduino-uno/doc.txt:
```diff
@@ -42,6 +42,119 @@ flash
+to press a button.
+
+In order to use on-chip debugging, the capacitor needs however to be
+disconnected from the reset pin. With the original Arduino Uno this can be done
```
@aabadie: I forgot this one :-/ That's why the debugging didn't work

maribu commented:
I should also add doc on how debugging with the Arduino Lenoardo can be done, as that board was merged since I opened this PR

maribu commented:
@benpicco: You do not happen to own an AVR Dragen, do you?

On the Microduino CoreRF the JTAG pin mapping is:

| JTAG | Pin | Pin Label |
|------|-----|-----------|
| TDI  | PF7 | A0        |
| TDO  | PF6 | A1        |
| TMS  | PF5 | A2        |
| TCK  | PF4 | A3        |

So, on-chip debugging should be possible there, too.

maribu commented:
Rebased to resolve merge conflics

maribu commented:
@benpicco: I added the doc for OnChip Debugging with the Microduino CoreRF. I didn't test, though!

benpicco commented:
I only have one of [these devices](https://de.aliexpress.com/item/32884596598.html) that works for flashing, but no advanced debugging I think.

maribu commented:
@benpicco: Why closing?

benpicco commented:
Sorry I had the page open on my phone, must have unlocked itself in my pocket 😳

maribu commented:
> Sorry I had the page open on my phone, must have unlocked itself in my pocket flushed

I think they call this "butt-dialing" in English :laughing:

roberthartung commented:
Okay, cross-read and will have a look at it asap! :)

maribu commented:
@roberthartung: Ping?

Do you habe an AVR Dragon or an Atmel ICE, so that you could test?

maribu commented:
@roberthartung did you have any chance to look at this?

@aabadie: I just rebased to also apply the de-duplication to the ATmega256RFR2-XPRO. Can you confirm that I didn't break anything for that board?

maribu commented:
I'll rebase on top of https://github.com/RIOT-OS/RIOT/pull/12729 in order to also cover the ATmega1284P. This PR thus should not be merged before that PR gets merged in order to not pull in unrelated stuff.

@roberthartung: Ping?

maribu commented:
Done. Debugging with the ATmega1284P works like a charm. JTAG debugging is much less pain compared to debugWIRE debugging, as used e.g. for the ATmega328P.

maribu commented:
Rebased on master. This PR no longer is waiting on the addition of the ATmega1284P "board".

@roberthartung: Daily ping ;-)

maribu commented:
@roberthartung: Daily ping.

(I somehow missed yesterday. Is there a bot for annoying reviewers?)

maribu commented:
@roberthartung: Daily ping :wink:

maribu commented:
@roberthartung: Daily ping :wink:

roberthartung commented:
I have a UNO somewhere that I can use for testing today. Will keep you posted @maribu

roberthartung commented:
Testing post-poned a day. I can't modify the UNOs I have here and I somehow can't find the SIP atmega1284p's...

roberthartung commented:
Ok. Found both an atmega328p and atmega1248p in PDIP. I use an atmel ice for all of the debugging.

I was able to flash the atmega1284p right away. No debugging yet.

I was able to enable the atmega328p fuses using ISP withthe command provided in doc.txt. However no more ISP (as expected), but also no debug wire is working. Any ideas @maribu ?

roberthartung commented:
For debugging I get `did not find any USB device "usb"` any idea? Do I have to specify the debugger?

roberthartung commented:
Ok, debugging works for my `atmega1284p` if I pass `-4` to avarice. Maybe we should check the PROGRAMMER if it needs more options.

maribu commented:
@roberthartung: I extended the doc and renamed the `DEBUGDEVICE` variable to `AVR_DEBUGDEVICE`. This will allow you to just export it from `.profile` or `.bashrc` according to the debugger you have. (So `export AVR_DEBUGDEVICE=--edbg` should enforce that your Atmel ICE is used on all AVR boards.)

@maribu: Note to myself: Don't forget to update the commit messages accordingly.

maribu commented:
Or may I squash right away to fix the commit messages?

maribu commented:
I just wanted to buy a second AVR Dragon and noticed that since I started this PR, the debugger has been dropped out of production. With that in mind, I changed the default debugger to the Atmel-ICE/JTAGICE3/EDBG.

@roberthartung: This means your debugger should now work out of the box, without having to change anything ;-)

roberthartung commented:
@maribu Hm is there any other chance to disable debugWire then, if AVR Dragons are not sold anymore? I wanted to buy one myself but no luck. And I have an 328p in debugWire mode here ;)

maribu commented:
The Atmel ICE claims to support debugWIRE. It is now actually a better buy, as it supports debugging of both Atmel's AVR and ARM chips. (But unlike the Dragon, no high voltage programming as far as I understood.) Sadly, the cheapest version of the Atmel ICE with no case and cables is still about twice as expensive as the Dragon was.

Another option for debugWIRE is this: https://github.com/dcwbrown/dwire-debug/blob/master/Manual.md

I tried this successfully, but it doesn't have the transparent breakpoint emulation features of the Dragon or the Atmel ICE. (But you can use inline assembly to instert a break instruction yourself, which works quite well. But that is rather static.)

maribu commented:
BTW: I used a cheap chinese USB-UART-Bridge for dwire-debug. So no need to buy and expensive FTDI for that.

roberthartung commented:
I have an original ICE, but can't get the debugWire disabled in order to re-program for now. Let me check again!

maribu commented:
Rebased (without squashing) to resolve merge conflicts.

Btw: I also ordered an Atmel ICE, so that I will be able to test if it works with that debugger as well. It might arrive in time for the holiday season :-)

maribu commented:
Btw: May I squash to fix the commit message? (It currently still says that the Dragon is the default, but I changed this as the AVR Dragon is no longer produced to the Atmel ICE, which now is the most capable and cheapest debugger for AVR devices.)

roberthartung commented:
Sure, go ahead to fix the commit message @maribu 

Let me test again so we are ready to build. Unfortunately I will only be able to test with the AtmelICE.

benpicco commented:
I would argue to just merge this.

 - it includes nice doc cleanups
 - there are no regressions when no debugger is used
 - it makes life much easier for people with a debugger.

So please rebase & squash.

maribu commented:
I rebased and updated the commit messages as now the Atmel-ICE rather than the (no longer sold) AVR Dragon is the default debugger. I also sneaked in some documentation updates to the Arduino Mega2560 (last commit) which contained a horribly outdated state tracking list. (All missing features there are already added to RIOT in the meantime.)

benpicco commented:
Travis has some nits.

maribu commented:
:tada: This makes debugging on ATmega boards much easier now :-) Thanks!
